pkg/rtc/transport.go

```diff
@@ -325,30 +335,43 @@ func (t *PCTransport) createAndSendOffer(options *webrtc.OfferOptions) error {
 	// if restart is requested, and we are not ready, then continue afterwards
 	if iceRestart {
 		if t.pc.ICEGatheringState() == webrtc.ICEGatheringStateGathering {
-			t.logger.Debugw("restart ICE after gathering")
+			t.params.Logger.Debugw("restart ICE after gathering")
 			t.restartAfterGathering = true
 			return nil
 		}
-		t.logger.Debugw("restarting ICE")
+		t.params.Logger.Debugw("restarting ICE")
 	}
 
-	// when there's an ongoing negotiation, let it finish and not disrupt its state
-	if t.negotiationState == negotiationStateClient {
+	if iceRestart && t.negotiationState != negotiationStateNone {
 		currentSD := t.pc.CurrentRemoteDescription()
-		if iceRestart && currentSD != nil {
-			t.logger.Debugw("recovering from client negotiation state")
+		if currentSD == nil {
+			// never received an answer from client, create a new peer connection
+			t.params.Logger.Infow("no previous answer from client on ICE restart, creating a new PC")
+			t.pc.Close()
+			if err := t.createPeerConnection(); err != nil {
+				t.params.Logger.Errorw("could not create new PC on ICE restart", err)
+				return err
+			} else {
+				options.ICERestart = false
+			}
```

Comment:
Would this work @cnderrauber @davidzhao ? i. e. server never got an answer from the client. So, server cannot recover by re-applying the remote description and moving forward. In that case, just closing the PC and re-opening a new one. Will that work?

Reply:
I think in the scenario that client didn't receive the offer, it might be ok? However, if the client did receive the offer, but we missed the answer, it will not be able to handle any of these sequences: [it doesn't have handling for recovery](https://github.com/livekit/client-sdk-js/blob/22ee04aec41174acc0163c23527b9f979d560c5d/src/room/PCTransport.ts#L31).

your change is fine.. I think we'll need to figure out how to detect the client being stuck and trigger a full reconnect.

Reply:
Yeah, will look at adding a monitor of sorts which can be more defensive and recover from stuck situations.

Reply:
@boks1971 it will not work here, some callback of pc is setup in the participant.go when transport created, and data channel created in there too, the new pc will not have this. And transport.pc is accessed by participant without lock protected, so create a new one might cause race issue.

Reply:
@cnderrauber Thank you. Good catch. Forgot about it. Thank you for looking into it.